practical_07/project_management.py

```diff
@@ -0,0 +1,207 @@
+    input_date = get_valid_date("Show projects that begin after date (dd/mm/yyyy): ")
+    projects.sort(key=attrgetter('start_date'))
+    for project in projects:
+        if input_date <= project.start_date:
+            print(project)
+
+
+def add_new_project(projects):
+    name = get_valid_string("Name: ")
+    start_date = get_valid_date("Start date (dd/mm/yyyy): ")
+    priority = get_valid_positive_int("Priority: ")
+    cost_estimate = float(input("Cost estimate: $"))
+    completion_percentage = get_valid_positive_int("Percent complete: ")
+    new_project = Project(str(name), start_date, int(priority), float(cost_estimate), int(completion_percentage))
+    projects.append(new_project)
+
+
+def update_project(projects):
+    for i, project in enumerate(projects):
+        print(f"{i} {project}")
+    project_number = get_valid_positive_int("Project choice: ")
+    while project_number > (len(projects) - 1):
+        print(f"Max project number = {len(projects) - 1}")
+        project_number = get_valid_positive_int("Project choice: ")
+    print(projects[project_number])
+    new_percentage = get_valid_positive_int_or_enter_skips("New percentage: ")
+    if new_percentage != "":
+        projects[project_number].completion_percentage = new_percentage
+    new_priority = get_valid_positive_int_or_enter_skips("New Priority: ")
+    if new_priority != "":
+        projects[project_number].priority = new_priority
+
+
+def get_valid_string(prompt):
+    item = input(prompt)
+    while item == "":
+        print("Invalid")
+        item = input(prompt)
+    return item
+
+
+def get_valid_date(prompt):
+    done = False
+    while not done:
+        input_date = input(prompt)
+        try:
+            input_date = datetime.strptime(input_date, '%d/%m/%Y').date()
+            done = True
+        except ValueError:
+            print("Invalid date format. Use dd/mm/yyyy format")
+    return input_date
+
+
+def get_valid_positive_int(prompt):
+    done = False
+    while not done:
+        try:
+            positive_int = int(input(prompt))
+            if positive_int < 0:
+                done = False
+                print("Input whole positive integers only")
+            else:
+                done = True
+        except ValueError:
+            print("Input whole positive integers only")
+    return positive_int  # Ignore error
+
+
+def get_valid_positive_int_or_enter_skips(prompt):
+    done = False
+    while not done:
+        positive_int = input(prompt)
+        if positive_int == "":
+            done = True
+        else:
+            try:
+                positive_int = int(positive_int)
+                if positive_int < 0:
+                    print("Input whole positive integers only")
+                else:
+                    done = True
+            except ValueError:
+                print("Input whole positive integers only")
+    return positive_int  # Ignore error
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
I didn't run this, however, if it works that is impressive as I struggled with this exercise in particular. You have a lot of added functions that I did not have.